### ISSUE TITLE
Installation error: undefined symbol: PyFPE_jbuf

### ISSUE THREAD

**Issue by datakid** [marcelm/cutadapt#196]
On an up to date Centos 7 installation, I can't do anything to make this error go away. I understand that it shouldn't effect our ability to run the software, but the error remains.

```
[root@slurm-01 cutadapt-1.9.1]# find . -name "*.pyc" -exec rm -rf {} \;
[root@slurm-01 cutadapt-1.9.1]# 
[root@slurm-01 cutadapt-1.9.1]# /usr/bin/python2.7 setup.py build_ext -i
running build_ext
[root@slurm-01 cutadapt-1.9.1]# bin/cutadapt --help

ERROR: A required extension module could not be imported because it is
incompatible with your system. A quick fix is to recompile the extension
modules with the following command:

    /usr/bin/python setup.py build_ext -i

See the documentation for alternative ways of installing the program.

The original error message follows.

Traceback (most recent call last):
  File "bin/cutadapt", line 9, in <module>
    from cutadapt.scripts import cutadapt
  File "/config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/scripts/cutadapt.py", line 62, in <module>
    check_importability()
  File "/config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/__init__.py", line 9, in check_importability
    import cutadapt._align
ImportError: /config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/_align.so: undefined symbol: PyFPE_jbuf
[root@slurm-01 cutadapt-1.9.1]# python2.7 bin/cutadapt --help

ERROR: A required extension module could not be imported because it is
incompatible with your system. A quick fix is to recompile the extension
modules with the following command:

    /usr/bin/python2.7 setup.py build_ext -i

See the documentation for alternative ways of installing the program.

The original error message follows.

Traceback (most recent call last):
  File "bin/cutadapt", line 9, in <module>
    from cutadapt.scripts import cutadapt
  File "/config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/scripts/cutadapt.py", line 62, in <module>
    check_importability()
  File "/config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/__init__.py", line 9, in check_importability
    import cutadapt._align
ImportError: /config/binaries/cutadapt/cutadapt-1.9.1/cutadapt/_align.so: undefined symbol: PyFPE_jbuf
[root@slurm-01 cutadapt-1.9.1]# 

```

**Comment by datakid** [comment]
I'm not pip-installing because it's being used on a slurm cluster and I want it on a space shared by all servers.

**Comment by marcelm** [comment]
We also have a SLURM cluster and I have written a cutadapt module (we use Lmod) for it. You can in fact pip install it. I just wrote up how I do it and added a section to the documentation, see <http://cutadapt.readthedocs.org/en/latest/installation.html#shared-installation-on-a-cluster>.

I recommend that approach, but if it doesn’t apply, then you can still go the setup.py route. I think the problem above is that the module was not actually recompiled when you ran the `build_ext` step. To make sure that it is, you need to delete the `cutadapt/*.so` files (the extension modules live in .so files, not in .pyc files). Note that this is not really a proper installation (that’s why it says "quick fix"). For that, you would need to run `python setup.py install` instead (with the proper options).

**Comment by datakid** [comment]
Ah! I ended up setting up with ```python setup.py install --prefix= ``` and ```python3 setup.py install --prefix= ```, then having two modulefiles. 

The important part in the modulefiles, to avoid this error 
```
from cutadapt.scripts import cutadapt
ImportError: No module named cutadapt.scripts
```
is to ```setenv PYTHONPATH=$path/lib64/python[x]/site-packages/``` and where necessary ```set-alias python '/usr/bin/python3'```

**Comment by marcelm** [comment]
Feel free to do what you think is best, but I do have to reply, if only fo the sake of others finding this thread.

Setting `PYTHONPATH` in module files is not a good idea. It will seem to work, but other software will fail in weird ways. The problem is that `PYTHONPATH` affects every Python binary that a user might run. It even affects both Python 2 and 3 although most certainly the modules in the path are version-specific. There are so many Python scripts out there that are part of bioinformatics packages. Setting `PYTHONPATH` affects all of them. Maybe some of them even come with their own copy of cutadapt, but by setting `PYTHONPATH`, you override that. The proper mechanism to solve this is to use a virtual environment.

Another advantage of using the method I suggested is that the module file does not have to depend on/load a `python` module. You just need to have the module loaded (if you have a python module) while you run the installation. And then there’s no need to create two module files either, just pick a version for your users.

**Comment by datakid** [comment]
Thanks marcelm, I can see your logic. My next question, though, is how do I load the relevant virtualenv via environment module? I found [this post](stackoverflow.com/questions/22885134/how-to-load-virtualenv-using-environmental-module-file-tcl-script) but it doesn't seem very solid.

**Comment by marcelm** [comment]
I received a message where you ask how to "load the virtualenv". I’m not sure why I’m not seing it here. Anyway: The nice thing is that the virtualenv does not need to be activated! And in fact, it *should* not be, otherwise you get all the weird problems I mentioned above and there isn’t a great advantage over the solution you already have.

The only thing that activating a virtual environment really does is that it prepends the `bin` directory within the directory where the virtual environment is located to your `$PATH`. You can just as well call the binaries in the `bin` directory directly. All the magic of virtual environments is actually part of Python itself - the `virtualenv` program just creates the proper directory structure so that the `python` binary recognizes that it is within a virtual environment, and it then adjusts its paths accordingly.

So the only thing that needs to be taken care of is that
* the module adds a directory containing the `cutadapt` script to the `$PATH`
* the `cutadapt` script has a shebang (`#!`) line at the top that starts a `python` binary that sits within the virtual environment.

Both of this is achieved by the instructions I added to the manual and which I linked to above (`pip install` sets the shebang line to the path to the current python interpreter).

So in short: Do not activate the virtualenv and everything will work!

**Comment by datakid** [comment]
Yes, I deleted my comment because I realised what I was thinking wrong
within moments of hitting send :) I'd spent so long using virtualenvwrapper
that I forgot what virtualenv provided.

It's all sorted, your solution makes significantly more sense, and is *so*
much cleaner.

Thanks

------
The most dangerous phrase in the language is, "We've always done it this
way."

- Grace Hopper

**Comment by marcelm** [comment]
Alright, and sorry for explaining what you already know ;-)

**Comment by brianpenghe** [comment]
Sorry. What is a good solution to this?

**Comment by brianpenghe** [comment]
Anyway I solved my problem by typing:

pip uninstall cutadapt

pip install --user --upgrade cutadapt

**Comment by marcelm** [comment]
You can try `pip3 install` instead to get a cutadapt with multi-core support. (Which is only available on Python 3.)
